docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -0,0 +1,198 @@
+
+IAVL trees are versioned, but a snapshot only contains the version relevant for the snapshot height. All historical versions are ignored.
+
+IAVL trees are insertion-order dependent, so key/value pairs must be stored in an appropriate insertion order to produce the same tree branching structure and thus the same Merkle hashes.
+
+A chunk corresponds to a subtree key range within an IAVL tree, in insertion order, along with the Merkle proof from the root of the subtree all the way up to the root of the multistore (including the Merkle proof for the surrounding multistore.
+
+```go
+struct SnapshotChunk {
+    Store   string            // Name (key) of IAVL store in outer MultiStore
+    Keys    [][]byte          // Snapshotted keys in insertion order
+    Values  [][]byte          // Snapshotted values corresponding to Keys
+    Proof   []merkle.ProofOp  // Merkle proof from subtree root to MultiStore root
+}
+```
+
+This chunk structure is believed to be sufficient to reconstruct an identical IAVL tree by applying separate chunks, but this may require the chunks to be ordered in a certain way; further research is needed.
+
+We do not use IAVL RangeProofs, since these include redundant data such as proofs for intermediate and leaf nodes that can be derived from the above data.
+
+Chunks should be built greedily by collecting key/value pairs constituting a complete subtree up to some size limit (e.g. 16 MB), then serialized and stored in the file system as `snapshots/<height>/<format>/<chunk>/data` and served via `RequestGetSnapshotChunk`.
+
+### Snapshot Scheduling
+
+Snapshots should be taken at some configurable height interval, e.g. every 1000 blocks. All nodes should preferably have the same snapshot schedule, such that all nodes can serve chunks for a given snapshot.
+
+Taking consistent snapshots of IAVL trees is greatly simplified by them being versioned: simply snapshot the version that corresponds to the snapshot height, while concurrent writes create new versions. IAVL pruning must not prune a version that is being snapshotted.
+
+Snapshots must also be garbage collected after some configurable time, e.g. by keeping the latest `n` snapshots.
+
+## Open Questions
+
+* Is it possible to reconstruct an identical IAVL tree given separate subtrees in an appropriate order, or is more data needed about the branch structure?
+
+* Should we punish nodes that provide invalid snapshots?
+
+* Should `ResponseOfferSnapshot` and `ResponseApplySnapshotChunk` return specific codes for e.g. invalid heights, unknown formats, and so on?
+
+* Is it OK for state-synced nodes to not have historical blocks nor historical IAVL versions?
+
+* Do we need to reject chunk sizes beyond a configurable limit, or will this be handled implicitly by e.g. gRPC request size limits?
```

Comment:
Explicit limit is better than implicit

Reply:
True. I'll add a 100 MB chunk size limit. However, since Gaia should use chunk compression, it also needs to be careful when decompressing, since an attacker could send heavily compressed chunks (e.g. 99% compression rate) and cause an OOM DoS.